server/graphql/v2/mutation/AddedFundsMutations.ts

```diff
@@ -0,0 +1,352 @@
+import { fetchOrderWithReference, GraphQLOrderReferenceInput, ObjectReference } from '../input/OrderReferenceInput';
+import { GraphQLTaxInput, TaxInput } from '../input/TaxInput';
+import { fetchTierWithReference, GraphQLTierReferenceInput } from '../input/TierReferenceInput';
+import {
+  fetchTransactionsImportRowWithReference,
+  GraphQLTransactionsImportRowReferenceInput,
+  GraphQLTransactionsImportRowReferenceInputFields,
+} from '../input/TransactionsImportRowReferenceInput';
+import { GraphQLOrder } from '../object/Order';
+
+type AddFundsMutationArgs = {
+  fromAccount: Record<string, unknown>;
+  account: Record<string, unknown>;
+  tier: Record<string, unknown>;
+  amount: AmountInputType;
+  paymentProcessorFee?: AmountInputType;
+  description: string;
+  memo: string;
+  processedAt: Date;
+  hostFeePercent: number;
+  invoiceTemplate: string;
+  tax: TaxInput;
+  accountingCategory: GraphQLAccountingCategoryReferenceInputFields;
+  transactionsImportRow: GraphQLTransactionsImportRowReferenceInputFields;
+};
+
+const validateAddFundsArgs = ({ account, fromAccount, args, paymentProcessorFee, totalAmount, host, req }) => {
+  const accountAllowedTypes = ['ORGANIZATION', 'COLLECTIVE', 'EVENT', 'FUND', 'PROJECT'];
+  if (!accountAllowedTypes.includes(account.type)) {
+    throw new ValidationFailed(
+      `Adding funds is only possible to the following types: ${accountAllowedTypes.join(',')}`,
+    );
+  }
+
+  if (account.isFrozen()) {
+    throw new ValidationFailed('Adding funds is not allowed for frozen accounts.');
+  }
+
+  // For now, we'll tolerate internal Added Funds whatever the type
+  // because we found it was a practice for Independent Collectives especially
+  const isInternal =
+    account.id === fromAccount.id ||
+    (account.ParentCollectiveId && account.ParentCollectiveId === fromAccount.id) ||
+    (fromAccount.ParentCollectiveId && account.id === fromAccount.ParentCollectiveId);
+  if (!isInternal) {
+    const fromAccountAllowedTypes = ['USER', 'ORGANIZATION', 'VENDOR'];
+    if (!fromAccountAllowedTypes.includes(fromAccount.type)) {
+      throw new ValidationFailed(
+        `Adding funds is only possible from the following types: ${fromAccountAllowedTypes.join(',')}`,
+      );
+    }
+  }
+
+  if (args.hostFeePercent < 0 || args.hostFeePercent > 100) {
+    throw new ValidationFailed('hostFeePercent should be a value between 0 and 100.');
+  } else if (args.tax && (args.tax.rate < 0 || args.tax.rate > 1)) {
+    throw new ValidationFailed('Tax rate must be between 0 and 1');
+  } else if (paymentProcessorFee > totalAmount) {
+    throw new ValidationFailed('Payment processor fee cannot be higher than the total amount');
+  } else if (args.hostFeePercent === 100 && paymentProcessorFee > 0) {
+    throw new ValidationFailed('Payment processor fee cannot be applied when host fee is 100%');
+  } else if (!totalAmount) {
+    throw new ValidationFailed('Amount should be greater than 0');
+  }
```

Comment:
This was covered by the add funds mutation, we need to make sure it's covered in the new one too

```suggestion
  } else if (!totalAmount || totalAmount < 0) {
    throw new ValidationFailed('Amount should be greater than 0');
  }
```